### PR TITLE
Apicurio registry operator

### PR DESCRIPTION
**Pull request by jsenko**
*Alpha* version of the operator:
- Creates *Deployment*, *Service*, *Ingress*
- Aligns image version and type
- Aligns replica count

TODO:
- Finish & verify persistence configuration alignment
- other configuration, i.e. CPU/Memory
- feature: autoscaling

See #18 for the old PR & comments.

### CODE REVIEW

**Comment by EricWittmann** [comment]
Hey @eguzki could you give this a look?  Lots of changes from @jsenko and would love another round of review!

File: apicurio-registry/deploy/operator.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+          env:
+            - name: WATCH_NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+            - name: POD_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            - name: OPERATOR_NAME
```
Comment (eguzki) [inline apicurio-registry/deploy/operator.yaml:30]: You might want to add env vars to provide all images that the operator deploys. For this specific case, the apicurio image.

File: apicurio-registry/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
+Apicurio Registry Operator
+===
+
+Requirements
+---
+* Docker
+* [go](https://github.com/golang/go) (1.13+, with `export GO111MODULE='on'`), and `$GOPATH` and `$GOROOT` set. 
+* [Operator SDK](https://github.com/operator-framework/operator-sdk/blob/master/doc/user/install-operator-sdk.md) v0.9.0+    
+* A running Kubernetes or [Minikube](https://kubernetes.io/docs/tasks/tools/install-minikube/) cluster,
```
Comment (eguzki) [inline apicurio-registry/README.md:9]: You might want to specify the minimum version of the kubernetes cluster supported.

File: apicurio-registry/pkg/controller/apicurioregistry/control_loop.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		if err != nil {
+			log.Error(err, "Error during the RESPOND phase of '"+v.Describe()+"' CF.")
+			requeue = true
+			continue
+		}
+	}
+
+	// Update the spec
+	spec = this.ctx.factory.CreateSpec(spec)
+	err = this.client.Update(context.TODO(), spec)
```
Comment (eguzki) [inline apicurio-registry/pkg/controller/apicurioregistry/control_loop.go]: Why update ALWAYS the spec? what is being updated every reconcilliation loop?

File: apicurio-registry/pkg/controller/apicurioregistry/control_loop.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	if err != nil {
+		log.Error(err, "Error updating spec")
+		return reconcile.Result{}, err
+	}
+	// Update the status
+	//spec = this.ctx.factory.CreateSpec(spec)
+	err = this.client.Status().Update(context.TODO(), spec)
```
Comment (eguzki) [inline apicurio-registry/pkg/controller/apicurioregistry/control_loop.go:124]: Same on status field. Usually the desired status is computed and if there are differences with the existing status, then updated. Usual reconcilliation pattern.

File: apicurio-registry/pkg/controller/apicurioregistry/deployment_cf.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+func (this *DeploymentCF) Describe() string {
+	return "Deployment Creation"
+}
+
+func (this *DeploymentCF) Sense(spec *ar.ApicurioRegistry, request reconcile.Request) error {
+	// Try to check if there is an existing deployment resource
+	deploymentName := this.ctx.configuration.GetConfig(CFG_STA_DEPLOYMENT_NAME)
+
+	deployments, err := this.ctx.kubecl.client.AppsV1().Deployments(this.ctx.configuration.GetSpecNamespace()).List(
```
Comment (eguzki) [inline apicurio-registry/pkg/controller/apicurioregistry/deployment_cf.go:42]: I do not get this. The operator expects to have multiple deployments per `ApicurioRegistry` instance?

If answer is not, you can request the deployment by specific name. The name can be some prefix and apicurioregistry name if multiple  `ApicurioRegistry` instances are allowed in a single namespace.

File: apicurio-registry/pkg/controller/apicurioregistry/control_loop.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	// GetConfig possible config errors
+	if e, s := this.ctx.configuration.GetErrors(); e {
+		err := errors.New(s)
+		log.Error(err, s)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// The LOOP
+	requeue := false
+	for _, v := range this.controlFunctions {
```
Comment (eguzki) [inline apicurio-registry/pkg/controller/apicurioregistry/control_loop.go:93]: The control loop is not clear enough for me. 

The reconcilliation pattern you should follow for each resource object the operator manages is (in pseudo code):
```
object = k8sclient.get(objectname)
if !exists(object) {
  k8sclient.create(object)
  return
}

if updateNeeded(object, desiredObject) {
  newObject = applyPatches(object, desiredObject)
  k8sclient.update(newObject)
}
```

For each object type, you may have specific implementation of those abstract methods like `updateNeeded` or `applyPatches`. For instance, the 3scale operator has some objects it does not reconcilliate, it just deploys them initially. So `updateNeeded` method always returns `false`.

One of the design decision might be being very restrictive and desiredObject must match exactly existing object. Another design decision might be only  reconciliate specific attributes of the object, for instance the replica count, which is directly affected by the `ApicurioRegistry` CR fields. This kind of decision are represented in the  implementation of that `updateNeeded` method.

File: apicurio-registry/pkg/controller/apicurioregistry/control_loop.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+type ApicurioRegistryReconciler struct {
+	client           client.Client
+	scheme           *runtime.Scheme
+	controlFunctions []ControlFunction
+	notInitialized   bool
+	ctx              *Context
+	controller       controller.Controller
+}
+
+func NewApicurioRegistryReconciler(mgr manager.Manager) *ApicurioRegistryReconciler {
```
Comment (eguzki) [inline apicurio-registry/pkg/controller/apicurioregistry/control_loop.go:26]: this file should be named `apicurioregistry_reconciler.go`

File: apicurio-registry/pkg/controller/apicurioregistry/kubecl.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+func newKubeClOrDie() kubernetes.Interface {
+	cfg, err := inClusterConfig()
+	if err != nil {
+		panic(err)
+	}
+	return kubernetes.NewForConfigOrDie(cfg)
+}
+
+func inClusterConfig() (*rest.Config, error) {
```
Comment (eguzki) [inline apicurio-registry/pkg/controller/apicurioregistry/kubecl.go:36]: This functionality is already given by the controller manager and you have available at ctx.client field.

**Review by eguzki — COMMENTED** [review]
Looking very good, congrats :1st_place_medal: 

I have dropped some comments, mainly some food for thought

cc @EricWittmann

**Review by EricWittmann — APPROVED** [review]

**Pull request merged**